internal/campaigns/service.go

```diff
@@ -199,19 +202,68 @@ func (svc *Service) NewExecutor(opts ExecutorOpts) Executor {
 	return newExecutor(opts, svc.client, svc.features)
 }
 
-func (svc *Service) NewWorkspaceCreator(dir string, cleanArchives bool) *WorkspaceCreator {
-	return &WorkspaceCreator{dir: dir, client: svc.client, deleteZips: cleanArchives}
+func (svc *Service) NewWorkspaceCreator(dir string, cleanArchives bool) WorkspaceCreator {
+	if svc.workspace == "volume" {
+		return &dockerVolumeWorkspaceCreator{client: svc.client}
+	}
+	return &dockerBindWorkspaceCreator{dir: dir, client: svc.client, deleteZips: cleanArchives}
+}
+
+// dockerImageSet represents a set of Docker images that need to be pulled. The
+// keys are the Docker image names; the values are a slice of pointers to
+// strings that should be set to the content digest of the image.
+type dockerImageSet map[string][]*string
+
+func (dis dockerImageSet) add(image string, digestPtr *string) {
+	if digestPtr == nil {
+		// Since we don't have a digest pointer here, we just need to ensure
+		// that the image exists at all in the map.
+		if _, ok := dis[image]; !ok {
+			dis[image] = []*string{}
+		}
+	} else {
+		// Either append the digest pointer to an existing map entry, or add a
+		// new entry if required.
+		if digests, ok := dis[image]; ok {
+			dis[image] = append(digests, digestPtr)
+		} else {
+			dis[image] = []*string{digestPtr}
+		}
+	}
```

Comment:
I get the feeling this was written by someone who's used to working with pointers 😄